### PR TITLE
[PT-D][Sharding] Enable more ops needed in the transformer model training

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #77214

From the code base of MetaSeq Model, we have found that loads of ops are not supported by sharded tensor. In https://github.com/pytorch/pytorch/pull/75374, we have enabled most of ops already and this PR/diff aims at enabling the rest of them.

Fix some unit test errors.

Differential Revision: [D36302780](https://our.internmc.facebook.com/intern/diff/D36302780/)

### CODE REVIEW

File: torch/distributed/_shard/sharded_tensor/_ops/math_ops.py [hunk 1]
```diff
@@ -83,21 +83,25 @@ def binary_math_op(types, args=(), kwargs=None, pg=None):
     # add
     torch.add,
     Tensor.add,
+    Tensor.add_,
```
Comment (wanchaol) [inline torch/distributed/_shard/sharded_tensor/_ops/math_ops.py:86]: is this working? I think we need to add a `add_` method to ShardedTensor class and dispatch to here to make it really work

File: test/distributed/_shard/sharded_tensor/ops/test_matrix_ops.py [hunk 2]
```diff
@@ -158,9 +132,6 @@ def _test_masked_fill_with_sizes(self, mask_size, broadcast_style=False):
                 mask = mask.unsqueeze(1)
             tensor_m = tensor.masked_fill(mask, 25.0)
             st_expected = _shard_tensor(tensor_m, spec)
-            st_expected._metadata.shards_metadata.sort(
-                key=lambda x: x.shard_offsets[0],
```
Comment (wanchaol) [inline test/distributed/_shard/sharded_tensor/ops/test_matrix_ops.py:162]: why we don't need sort this time?

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/_common.py [hunk 3]
```diff
@@ -57,7 +57,7 @@ def _register_sharded_op_on_local_tensor(
     """
     @sharded_op_impl(op)
     @_sharded_op_common(op, early_stop_func, extra_check)
-    def sharded_tensor_op_on_local_shards(types, args=(), kwargs=None, pg=None):
+    def sharded_tensor_op_on_local_tensor(types, args=(), kwargs=None, pg=None):
```
Comment (wanchaol) [inline torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/_common.py:60]: hmmm we didn't change the callsites of `_register_sharded_op_on_local_shards` to `_register_sharded_op_on_local_tensor`, is this already been changes somewhere else?

File: test/distributed/_shard/sharded_tensor/ops/test_matrix_ops.py [hunk 4]
```diff
@@ -132,23 +123,6 @@ def test_sharded_tensor_softmax(self):
                 )
             )
 
-    @with_comms(init_rpc=True)
-    @skip_if_lt_x_gpu(TEST_GPU_NUM)
-    @requires_nccl()
-    def test_sharded_tensor_softmax_error(self):
```
Comment (wanchaol) [inline test/distributed/_shard/sharded_tensor/ops/test_matrix_ops.py:138]: are we supporting softmax on sharding dim already, why we delete this test?

File: torch/distributed/_shard/sharded_tensor/_ops/math_ops.py (hunk 1 above)
Comment (fduwjj) [reply torch/distributed/_shard/sharded_tensor/_ops/math_ops.py:86]: Hmm, that's wired because the unit test does pass.

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/_common.py (hunk 3 above)
Comment (fduwjj) [reply torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/_common.py:60]: Nope, we don't need to change the callsite because this is the func inside the decorator. And we are not changing the signature of decorator itself.

File: test/distributed/_shard/sharded_tensor/ops/test_matrix_ops.py (hunk 2 above)
Comment (fduwjj) [reply test/distributed/_shard/sharded_tensor/ops/test_matrix_ops.py:162]: Because your PR(https://github.com/pytorch/pytorch/pull/75991) has fixed it so we now don't need it anymore.

**Review by wanchaol — APPROVED** [review]
looks good to me! Thanks for fixing

File: torch/distributed/_shard/sharded_tensor/_ops/matrix_ops.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+    Return: None
+    """
+    if len(args) < 2:
+        raise ValueError("Needs to give a tensor to cast type as!")
+    if not isinstance(args[1], torch.Tensor) and not isinstance(args[1], ShardedTensor):
+        raise ValueError("Needs to give a Tensor or ShardedTensor to cast type as!")
+
+
+def same_dtype(*args, **kwargs):
```
Comment (pritamdamania87) [inline torch/distributed/_shard/sharded_tensor/_ops/matrix_ops.py:29]: Weren't these already added in https://github.com/pytorch/pytorch/pull/75374?

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/softmax.py [hunk 6]
```diff
@@ -14,14 +14,14 @@ def sharded_softmax(types, args=(), kwargs=None):
...
-    ndims = len(input.size())
+    ndims = input.dim()
     if dim == sharding_dim or dim + ndims == sharding_dim or sharding_dim + ndims == dim:
         exp = torch.exp(input.local_tensor())
         exp_sum = exp.sum(dim=dim).unsqueeze(dim=dim)
         exp_sum = torch.distributed.nn.functional.all_reduce(exp_sum, group=pg)
         smax = torch.div(exp, exp_sum)
     else:
-        smax = torch.nn.functional.softmax(input.local_tensor())
+        smax = torch.nn.functional.softmax(input.local_tensor(), dim=dim)
```
Comment (pritamdamania87) [inline torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/softmax.py:24]: Thanks for fixing this! Can we extend `test_softmax.py` to ensure there was a unit test covering this?

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/matrix_ops.py [hunk 7]
```diff
@@ -304,3 +311,102 @@ def sharded_bmm(args, kwargs, pg):
...
+    st = args[0]
+    normalized_shape = args[1]
+    local_tensor = st.local_tensor()
+    current_rank = dist.get_rank(pg)  # type: ignore[attr-defined]
+    global_tensor = (
+        torch.empty(st.size(), device=st.local_tensor().device)
+        if current_rank == 0
+        else None
+    )
+    st.gather(dst=0, out=global_tensor)
```
Comment (pritamdamania87) [inline torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/matrix_ops.py:363]: This seems pretty inefficient, can't we only do one allreduce to compute the `mean` and then given the `mean` we can compute the `standard deviation` as well and do rest of the operations just locally?

**Review by pritamdamania87 — CHANGES_REQUESTED** [review]

File: torch/distributed/_shard/sharded_tensor/_ops/matrix_ops.py (hunk 5 above)
Comment (fduwjj) [reply torch/distributed/_shard/sharded_tensor/_ops/matrix_ops.py:29]: Yep. It somehow gets removed during rebase? I also felt wired about `type_as`. Originally the test fails so I have to add them back.

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/softmax.py (hunk 6 above)
Comment (fduwjj) [reply torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/softmax.py:24]: Sure, maybe I can move this into `test_softmax.py`?

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/matrix_ops.py (hunk 7 above)
Comment (fduwjj) [reply torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/matrix_ops.py:363]: I see. sure, let me change it like that.

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/softmax.py (hunk 6 above)
Comment (fduwjj) [reply torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/softmax.py:24]: Looks like `test_softmax.py` already contains this but it somehow use the default dim as the sharding dim so we didn't find out the corner case.

**Comment by fduwjj** [comment]
Rebase and addressed the comment from reviewers.

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/matrix_ops.py [hunk 8]
```diff
@@ -304,3 +313,100 @@ def sharded_bmm(args, kwargs, pg):
...
+
+    norm_dims = tuple(i for i in range(-1, -len(normalized_shape) - 1, -1))
+    local_size = math.prod(local_tensor.size()[shape_start:])  # type: ignore[attr-defined]
+    st_size = math.prod(st.size()[shape_start:])  # type: ignore[attr-defined]
+    local_mean = torch.mul(local_tensor.mean(norm_dims, keepdim=True), local_size)
+    global_mean = torch.div(all_reduce(local_mean), st_size)
+    local_variant_sq = torch.square(local_tensor - global_mean).sum(
+        norm_dims, keepdim=True
+    )
+    global_variant = torch.div(all_reduce(local_variant_sq), st_size)
```
Comment (pritamdamania87) [inline torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/matrix_ops.py:380]: Do we need this additional allreduce? Shouldn't the global mean be sufficient to compute the global variance?

File: test/distributed/_shard/sharded_tensor/ops/test_matrix_ops.py [hunk 9]
```diff
@@ -258,6 +223,62 @@ def test_sharded_tensor_view_error(self):
...
                 st.view(5, 7, -1, -1)
 
+    @with_comms(init_rpc=True)
+    @skip_if_lt_x_gpu(TEST_GPU_NUM)
+    @requires_nccl()
+    def test_sharded_tensor_layer_norm(self):
+        specs = _chunk_sharding_specs_list_for_test([1, 2], seed=10)
+        for spec in specs:
+            tensor = torch.rand(16, 35, 26).cuda(self.rank)
+            layer_norm = torch.nn.LayerNorm((35, 26)).cuda(self.rank)
```
Comment (pritamdamania87) [inline test/distributed/_shard/sharded_tensor/ops/test_matrix_ops.py:233]: Can we add a test with `elementwise_affine`?

File: test/distributed/_shard/sharded_tensor/ops/test_softmax.py [hunk 10]
```diff
@@ -33,11 +33,11 @@ def _test_sharded_softmax(self, dim):
         local_tensor = torch.rand(10, 10, device=self.rank)
         local_softmax = torch.nn.functional.softmax(local_tensor, dim)
 
-        spec = ChunkShardingSpec(dim=0, placements=[f'rank:{idx}/cuda:{idx}' for idx in range(self.world_size)])
+        spec = ChunkShardingSpec(dim=1, placements=[f'rank:{idx}/cuda:{idx}' for idx in range(self.world_size)])
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply test/distributed/_shard/sharded_tensor/ops/test_softmax.py:36]: sure!

File: test/distributed/_shard/sharded_tensor/ops/test_matrix_ops.py (hunk 9 above)
Comment (fduwjj) [reply test/distributed/_shard/sharded_tensor/ops/test_matrix_ops.py:233]: Sure.

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/matrix_ops.py (hunk 8 above)
Comment (fduwjj) [reply torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/matrix_ops.py:380]: I believe so, the way we calculate global variance still needs information from all shards. Unless there is some fancy math which can help us avoid the extra communication. Correct me if I am wrong.

**Review by pritamdamania87 — APPROVED** [review]
